src/api/macros/chrysalis-macros.js

```diff
@@ -249,15 +218,64 @@ class Macros {
       }
     }
     return macros;
-  }
+  };
 
-  async focus(s, macros) {
+  const expandMacro = (macro) => {
+    let m = macro.map((v) => Object.assign({}, v));
+    const mapping = {
+      KEYCODEUP: "KEYUP",
+      KEYCODEDOWN: "KEYDOWN",
+      TAPCODE: "TAP",
+      TAPCODESEQUENCE: "TAPSEQUENCE",
+    };
+
+    for (let step of m) {
+      if (mapping[step.type]) {
+        step.type = mapping[step.type];
+      }
+    }
+
+    return m;
+  };
+
+  const clone = (macro) => {
+    return macro.map((v) => Object.assign({}, v));
+  };
+
+  const compressMacro = (macro) => {
+    let m = clone(macro);
+    for (let step of m) {
+      if (step.type == "KEYUP" && step.value.code < 256)
+        step.type = "KEYCODEUP";
+      if (step.type == "KEYDOWN" && step.value.code < 256)
+        step.type = "KEYCODEDOWN";
+      if (step.type == "TAP" && step.value.code < 256) step.type = "TAPCODE";
+      if (step.type == "TAPSEQUENCE") {
+        if (step.value.filter((c) => c.code >= 256).length == 0) {
+          step.type = "TAPCODESEQUENCE";
+        }
+      }
+    }
+    return m;
+  };
+
+  this.compress = (macros) => ({
+    storageSize: macros.storageSize,
+    macros: macros.macros.map((macro) => compressMacro(clone(macro))),
+  });
+
+  this.getStoredSize = (macros) =>
+    this.serialize(this.compress(macros).macros).length;
+
+  this.focus = async (s, macros) => {
     if (macros) {
-      const ser = this.serializeMacros(macros.macros);
-      if (ser.length > macros.storageSize) {
+      const m = this.compress(macros);
+      const ser = this.serialize(m.macros);
+      if (ser.length > m.storageSize) {
```

Comment:
Eventually, we should have a friendlier error message. At least "Your macros take x bytes, but your device only has y bytes for macro storage. "

Reply:
The message here isn't really interesting. There's `getStoredSize()` which the UI will use to check the size, and will disallow saving in the first place, so the UI will never hit this path. It's only there for development and possible reuse outside of the UI itself.

Reply:
Did that get sorted?

Reply:
Yes, the "Save changes" button on the main editor page is disabled when the macro space is overused, and a warning is displayed:

![Screenshot from 2022-05-27 22-24-37](https://user-images.githubusercontent.com/17243/170784672-50ae5ca3-3dba-4179-aa7e-e19ca74719a8.png)